docs/sdk/instructlab-sdk.md

```diff
@@ -0,0 +1,171 @@
+instructlab.core.Data.Generator
+instructlab.core.Data.Generator.generate_data
+instructlab.core.Model.Server.serve
+instructlab.core.Model.Trainer
+instructlab.core.Model.Trainer.train_model
+instructlab.core.Model.Trainer.process_data
+instructlab.core.Model.Evaluator
+instructlab.core.Model.Evaluator.mt_bench
+instructlab.core.Model.Evaluator.dk_bench
+instructlab.core.Model.Evaluator.mmlu_bench
+instructlab.core.RAG.Ingestor.ingest
+instructlab.core.RAG.Converter.convert
+```
+
+Presumably, the distinct methods under each class will grow, which is why I am opting to make very distinct classes per operation.
+
+These initial exposed functions can expand to include any new functionality that is more SDK oriented from the various libraries. For example, if SDG adds something like subset selection, teacher as annotator, data mixing, etc we could expose an `instructlab.core.Data.Annotator.annotate` or `instructlab.core.Data.Mixer.mix` that could be invoked in sequence in a user's script with other parts of the ilab workflow. Some things make _less_ sense to be exposed via a CLI, but still are critical to ensuring users get a good model and properly generated data.
+
+There are certain things that only exist in `ilab` currently and functionality that is going to be moving these such as data ingestion, RAG, etc. Forming an SDK for `instructlab` allows us to capture all of these concerns under one API.
+
+These endpoints in combination with the curated InstructLab Config File will open up these workflows to users and allow InstructLab to be easily incorporated into other projects. Allowing people to run things like data generation, and full fine-tuning via an SDK that pulls in their pre-existing config.yaml but also can be run independently will open new avenues for InstructLab adoption and extensibility.
+
+## Changes to the CLI
+
+The `ilab` CLI will need to adapt to this new structure. Commands like `ilab data generate` should, in terms of code, follow this flow:
+
+1. `src/instructlab/cli/data/generate.py`
+2. `src/instructlab/data/generate.py`
+3. `src/instructlab/process.py`
+4. `src/instructlab/core/data/generator.py`
+
+So generally: cli -> internal handling package -> process management package -> core SDK -> library code, is the flow.
+
+The flow of the CLI today is such that the cli package for a command (`src/instructlab/cli/data/generate.py`) parses the command line options, and passes control to the core code (`/src/instructlab/data/generate.py`). This then (if applicable), yields control to the `process` package which kicks off necessary processes for the given command.
+
+The difference with an SDK is that we would eventually want to end up executing `core/data/generator.py`, the actual publicly consumable python SDK. This will ensure that the CLI can do whatever custom handling it needs to do on top, but eventually it must boil down to the `core` package which uses publicly available methods from the various libraries.
```

Comment:
Would it not make more sense for both the CLI and "Core" to be consuming from the generation and process logic separately? Why couple the CLI with the SDK at all?

Reply:
It's possible we'd want this in some situations, but I think it's a good design decision to plop the CLI on top of the SDK as much as possible. For one thing, it minimizes duplicate code doing similar things. Another good thing is that we'd be testing our own SDK by wrapping it- testing the CLI would also be (only slightly indirectly) testing the SDK.

Reply:
I don't think core should consume process management at all. Vice versa, CLI - through process management layer, if needed - will consume SDK. The user of the SDK should not be required to use our particular process management model. (Nor I think process management module is interesting to general consumers per se as a separate library. Prefer keeping it private to ilab CLI flow.)

Reply:
Hmmm it's an interesting point @JamesKunstle @booxter - but isn't the difference here that the CLI and the SDK are consuming the same logic but exposing it in different ways, i.e. via the command line versus importable classes? I suppose the CLI could utilize the SDK classes, but at that point it seems to me that the logic itself could simply expose importable classes and that would cover the usecase? 

Indeed the process management piece shapes a lot of this conversation

Reply:
As far as I understood so far, process management is applicable where we anticipate long running processes (generate, train etc.) Some commands are like that and will go through the chain of calls as:

```
click -> (cli handler) - ...schedules process... -> process management -> ...spawns process... -> (process handler) -> sdk -> libraries
```

while for processes that don't need background execution, it would be:


```
click -> (cli handler) -> sdk -> libraries
```

Reply:
the SDK should be separate from the process logic, having bespoke state and process tracking through and SDK is out of the ordinary because a user might want to coordinate the processes themselves in a script. The current state of the document represents how I believe the flow should work